File: enterprise-starter-kit/src/main/java/com/enterprise/starter/kit/modules/customers/service/CustomerService.java
package com.enterprise.starter.kit.modules.customers.service;

import com.enterprise.starter.kit.modules.customers.dto.CustomerRequest;
import com.enterprise.starter.kit.modules.customers.dto.CustomerResponse;
import com.enterprise.starter.kit.modules.customers.entity.Customer;
import com.enterprise.starter.kit.modules.customers.repository.CustomerRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/// Service for customer management operations.
@Service
public class CustomerService {

    private final CustomerRepository customerRepository;
    private final com.enterprise.starter.kit.modules.tenant.repository.CompanyRepository companyRepository;

    public CustomerService(CustomerRepository customerRepository,
            com.enterprise.starter.kit.modules.tenant.repository.CompanyRepository companyRepository) {
        this.customerRepository = customerRepository;
        this.companyRepository = companyRepository;
    }

    @Transactional(readOnly = true)
    public Page<CustomerResponse> getAllCustomers(Pageable pageable) {
        return customerRepository.findAll(pageable).map(CustomerResponse::fromEntity);
    }

    @Transactional(readOnly = true)
    public Page<CustomerResponse> searchCustomers(String search, Pageable pageable) {
        return customerRepository.searchCustomers(search, pageable).map(CustomerResponse::fromEntity);
    }

    @Transactional(readOnly = true)
    public CustomerResponse getCustomerById(Long id) {
        Customer customer = customerRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Customer not found with id: " + id));
        return CustomerResponse.fromEntity(customer);
    }

    @Transactional(readOnly = true)
    public List<CustomerResponse> getActiveCustomers() {
        return customerRepository.findByIsActiveTrue().stream()
                .map(CustomerResponse::fromEntity)
                .toList();
    }

    /// Create a new customer
    @Transactional
    public CustomerResponse createCustomer(CustomerRequest request) {
        if (request.email() != null && customerRepository.existsByEmail(request.email())) {
            throw new IllegalArgumentException("Customer with this email already exists");
        }

        String tenantId = com.enterprise.starter.kit.shared.tenant.TenantContext.getTenantId();

        Customer customer = new Customer();
        mapRequestToEntity(request, customer);
        customer = customerRepository.save(customer);
        return CustomerResponse.fromEntity(customer);
    }

    /// Update an existing customer
    @Transactional
    public CustomerResponse updateCustomer(Long id, CustomerRequest request) {
        Customer customer = customerRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Customer not found with id: " + id));

        // Check email uniqueness if changed
        if (request.email() != null && !request.email().equals(customer.getEmail())) {
            if (customerRepository.existsByEmail(request.email())) {
                throw new IllegalArgumentException("Customer with this email already exists");
            }
        }

        mapRequestToEntity(request, customer);
        customer = customerRepository.save(customer);
        return CustomerResponse.fromEntity(customer);
    }

    /// Soft delete a customer (set inactive)
    @Transactional
    public void deactivateCustomer(Long id) {
        Customer customer = customerRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Customer not found with id: " + id));
        customer.setIsActive(false);
        customerRepository.save(customer);
    }

    /// Hard delete a customer
    @Transactional
    public void deleteCustomer(Long id) {
        if (!customerRepository.existsById(id)) {
            throw new IllegalArgumentException("Customer not found with id: " + id);
        }
        customerRepository.deleteById(id);
    }

    /// Count active customers
    public long countActiveCustomers() {
        return customerRepository.countByIsActiveTrue();
    }

    /// Bulk delete customers by IDs
    @Transactional
    public void bulkDelete(List<Long> ids) {
        if (ids == null || ids.isEmpty())
            return;
        customerRepository.deleteAllById(ids);
    }

    private void mapRequestToEntity(CustomerRequest request, Customer customer) {
        customer.setName(request.name());
        customer.setEmail(request.email());
        customer.setPhone(request.phone());
        customer.setAddress(request.address());
        customer.setCity(request.city());
        customer.setCountry(request.country());
        customer.setTaxNumber(request.taxNumber());
        customer.setContactPerson(request.contactPerson());
        customer.setNotes(request.notes());
    }
}

File: enterprise-starter-kit-frontend/src/app/services/customer.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface Customer {
  id: number;
  name: string;
  email: string;
  phone: string;
  address: string;
  city: string;
  country: string;
  taxNumber: string;
  contactPerson: string;
  notes: string;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface CustomerRequest {
  name: string;
  email?: string;
  phone?: string;
  address?: string;
  city?: string;
  country?: string;
  taxNumber?: string;
  contactPerson?: string;
  notes?: string;
}

export interface PageResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

@Injectable({
  providedIn: 'root'
})
export class CustomerService {
  private readonly apiUrl = `${environment.apiUrl}/customers`;
  private http = inject(HttpClient);

  getCustomers(page = 0, size = 20, sort = 'name'): Observable<PageResponse<Customer>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', sort);
    return this.http.get<PageResponse<Customer>>(this.apiUrl, { params });
  }

  searchCustomers(query: string, page = 0, size = 20): Observable<PageResponse<Customer>> {
    const params = new HttpParams()
      .set('q', query)
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<PageResponse<Customer>>(`${this.apiUrl}/search`, { params });
  }

  getActiveCustomers(): Observable<Customer[]> {
    return this.http.get<Customer[]>(`${this.apiUrl}/active`);
  }

  getCustomerById(id: number): Observable<Customer> {
    return this.http.get<Customer>(`${this.apiUrl}/${id}`);
  }

  createCustomer(customer: CustomerRequest): Observable<Customer> {
    return this.http.post<Customer>(this.apiUrl, customer);
  }

  updateCustomer(id: number, customer: CustomerRequest): Observable<Customer> {
    return this.http.put<Customer>(`${this.apiUrl}/${id}`, customer);
  }

  deleteCustomer(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  bulkDelete(ids: number[]): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/bulk`, { body: ids });
  }

  deactivateCustomer(id: number): Observable<void> {
    return this.http.patch<void>(`${this.apiUrl}/${id}/deactivate`, {});
  }

  countActiveCustomers(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/count`);
  }
}
